### DISCUSSION TITLE
Help to port Socket.IO to Emscripten (WebAssembly)

### DISCUSSION THREAD

**Discussion by skhaz** [socketio/socket.io#5251] [Q&A] [answered]
I am creating a Socket.IO implementation for Emscripten in C++, purely using WebSockets.

I followed the documentation for several days and even tried reverse-engineering the client for web browsers, but without success.

Below is a step-by-step description of what I am doing, and I would be very grateful to know what I did wrong to experience frequent disconnections.

**WebSocket URL** `ws://localhost:3000/socket.io/?transport=websocket&EIO=4` (no HTTP!)

I connect and I receive

```
0{"sid":"w4NlzSj3BPm24GGDAAAX","upgrades":[],"pingInterval":25000,"pingTimeout":20000,"maxPayload":1000000}
```

Then I anwser with

```
0
```

Then the connection keeps for while...

If I try to send

```
42["messsage","hello world"]
```

I got disconnected immediately.

Also, if I receive 

```
2
```

And anwser with

```
3
```

I got disconnected immediately too.

**Comment by darrachequesne** [comment, marked as answer]
Hi!

The protocol is split in two:

- the Engine.IO protocol
  - `0{"sid":"w4Nlz...}` is the handshake sent by the server
  - `2` are indeed PING packets, and the client should respond with a PONG packet (`3`)

Reference: https://socket.io/docs/v4/engine-io-protocol/

- the Socket.IO protocol

Every Socket.IO packet must be prefixed by `4`, indicating an Engine.IO MESSAGE packet.

So sending a CONNECT packet (`0`) becomes a `40` over the wire.

Then you will be able to send message with `42[...]`.

Reference: https://socket.io/docs/v4/socket-io-protocol/#format

**Comment by skhaz** [comment]
I see. Thank you!
